garglk/ttsmac.m

```diff
@@ -0,0 +1,155 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
+ * GNU General Public License for more details.                               *
+ *                                                                            *
+ * You should have received a copy of the GNU General Public License          *
+ * along with Gargoyle; if not, write to the Free Software                    *
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA *
+ *                                                                            *
+ *****************************************************************************/
+
+#import <Cocoa/Cocoa.h>
+
+#include "glk.h"
+#include "garglk.h"
+
+#import "sysmac.h"
+
+#define TXTSIZE 4096
+static glui32 txtbuf[TXTSIZE + 1];
+static size_t txtlen;
+
+static NSMutableString * ttstext = nil;
+static NSSpeechSynthesizer * synth = nil;
+
+void gli_initialize_tts(void)
+{
+    if (gli_conf_speak)
+    {
+        synth = [NSSpeechSynthesizer new];
+        if (!synth)
+        {
+            NSLog(@"Unable to initialize TTS support!");
+            return;
+        }
+
+        ttstext = [NSMutableString stringWithCapacity: TXTSIZE];
+
+        if (gli_conf_speak_language != NULL)
+        {
+            NSString * lang = [NSString stringWithCString: gli_conf_speak_language
+                                                 encoding: NSUTF8StringEncoding];
+
+            NSArray<NSString *> * voices = [NSSpeechSynthesizer availableVoices];
+            for (NSString * voice in voices)
+            {
+                NSDictionary<NSString *, id> * attr = [NSSpeechSynthesizer attributesForVoice: voice];
+                if ([lang isEqualToString: [attr objectForKey: NSVoiceLocaleIdentifier]])
+                {
+                    [synth setVoice: voice];
+                }
+            }
+        }
+
+        NSLog(@"TTS configured with voice: %@", [synth voice]);
+    }
+
+    txtlen = 0;
+}
+
+void gli_tts_flush(void)
+{
+    if (!synth)
+        return;
+
+    if ([ttstext length] > 0)
+    {
+        BOOL result = [synth startSpeakingString: ttstext];
+        if (result == YES)
+        {
+            [ttstext setString: @""];
+        }
+    }
+
+    txtlen = 0;
+}
+
+void gli_tts_purge(void)
+{
+    if (synth)
+    {
+        [synth stopSpeaking];
+    }
+}
+
+static void txtbuf_append(void)
+{
+    if (txtlen > 0)
+    {
+        NSString * text = [[NSString alloc] initWithBytes: txtbuf
+                                                   length: (txtlen * sizeof(glui32))
+                                                 encoding: UTF32StringEncoding];
+
+        if (text)
+        {
+            [ttstext appendString: text];
+            [text release];
+        }
+
+        txtlen = 0;
+    }
+}
+
+void gli_tts_speak(const glui32 *buf, size_t len)
```

Comment:
You are correct that nothing is spoken until `gli_tts_flush()` is called. Calling `[NSSpeechSynthesizer startSpeakingString:]` doesn't queue the text but immediately cancels whatever is being spoken and starts speaking the new message; calling it on a character by character basis wouldn't work. So, I opted to buffer everything until the next input prompt and then output all the text in a single call.

The tests I ran with a few games seems that it works relatively OK in practice (I wouldn't have made the PR otherwise). The macOS speech synthesizer is aware of paragraph breaks and pauses appropriately. This is one of the reasons I asked for feedback especially if someone could compare how it behaves against Linux or Windows builds.